lib/ddtrace/propagation/http_propagator.rb

```diff
@@ -11,26 +15,65 @@ module HTTPPropagator
     def self.inject!(context, env)
       # Prevent propagation from being attempted if context provided is nil.
       if context.nil?
-        Datadog::Tracer.log.debug('Cannot inject context into env to propagate over HTTP: context is nil.'.freeze)
+        ::Datadog::Tracer.log.debug('Cannot inject context into env to propagate over HTTP: context is nil.'.freeze)
         return
       end
 
-      env[HTTP_HEADER_TRACE_ID] = context.trace_id.to_s
-      env[HTTP_HEADER_PARENT_ID] = context.span_id.to_s
-      env[HTTP_HEADER_SAMPLING_PRIORITY] = context.sampling_priority.to_s
-      env[HTTP_HEADER_ORIGIN] = context.origin.to_s if context.origin
-      env.delete(HTTP_HEADER_SAMPLING_PRIORITY) unless context.sampling_priority
+      # Check if they want to propagate with B3 headers
+      if ::Datadog.configuration.propagation_inject_style.include?(PROPAGATION_STYLE_B3)
+        DistributedHeaders::B3.inject!(context, env)
+      end
+
+      # Check if they want to propagate with B3 single header
+      if ::Datadog.configuration.propagation_inject_style.include?(PROPAGATION_STYLE_B3_SINGLE_HEADER)
+        DistributedHeaders::B3Single.inject!(context, env)
+      end
+
+      # Check if they want to propagate with Datadog headers
+      if ::Datadog.configuration.propagation_inject_style.include?(PROPAGATION_STYLE_DATADOG)
+        DistributedHeaders::Datadog.inject!(context, env)
+      end
     end
 
     # extract returns a context containing the span ID, trace ID and
     # sampling priority defined in env.
     def self.extract(env)
-      headers = DistributedHeaders.new(env)
-      return Datadog::Context.new unless headers.valid?
-      Datadog::Context.new(trace_id: headers.trace_id,
-                           span_id: headers.parent_id,
-                           sampling_priority: headers.sampling_priority,
-                           origin: headers.origin)
+      context = nil
+      dd_context = nil
+      ::Datadog.configuration.propagation_extract_style.each do |style|
+        extracted_context = nil
+        case style
+        when PROPAGATION_STYLE_DATADOG
```

Comment:
Same though as to being more dynamic as stated above. (With the caveat that it might be reasonable to have an exception or two for something like `Datadog` propagation.)